pkg/kube/kclient/crdwatcher_test.go

```diff
@@ -26,6 +26,30 @@ import (
 	"istio.io/istio/pkg/test/util/assert"
 )
 
+// TestCRDWatcherRace tests for a previous bug where callbacks may be skipped if added during a handler
+func TestCRDWatcherRace(t *testing.T) {
+	stop := test.NewStop(t)
+	c := kube.NewFakeClient()
+	ctl := c.CrdWatcher()
+	go ctl.Run(stop)
+	vsCalls := atomic.NewInt32(0)
+
+	// Race callback and CRD creation
+	go func() {
+		if ctl.KnownOrCallback(gvr.VirtualService, func(s <-chan struct{}) {
+			assert.Equal(t, s, stop)
+			// Happened async
+			vsCalls.Inc()
+		}) {
+			// Happened sync
+			vsCalls.Inc()
```

Comment:
To be clear, this never gets called, correct?

Reply:
It can happen, since this test is intentionally causing a race its inconsistent. It happens if we make the CRD and  process it fully before we call KnownOrCallback.